<a href="https://colab.research.google.com/github/jimhaines37/DataScience/blob/main/_labs/Lab11/Lab11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 11: Scraping

This lab will walk you through the basics of using both the `requests` library as well as [Beautiful Soup]() to communicate with and retrieve data from websites on the internet.

**Instructions:** When you have completed the lab, please submit it via [Tulane Canvas](https://tulane.instructure.com/) before the due date.

In [1]:
# first, mount your google drive, change to the course folder, pull latest changes, and change to the lab folder.
from google.colab import drive
drive.mount('/content/drive')
%cd /content/drive/My Drive/cmps3160
!git pull
%cd _labs/Lab11

Mounted at /content/drive
/content/drive/My Drive/cmps3160
Already up to date.
/content/drive/My Drive/cmps3160/_labs/Lab11


## Scraping Data From the Internet

What happens when you type a URL into the search bar? Well, you can follow along at [this in depth guide](https://github.com/alex/what-happens-when) which takes you through all the layers of the IP stack, tracing every bit of data exchange. However, for our purposes we just need to understand a few basics:
1. When you hit `enter` on your web-browser, the browser issues a `GET` command to the webserver, asking for the default homepage, typically a page called `index.html`.
2. The server then sends back either the contents of that page with a HTTP code `200`, meaning everything worked, or it sends back an error like `404`, meaning that you asked for a page it doesn't know how to find. You can read more about [HTTP Error codes here](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).
3. Your browser then takes that page, perhaps issuing even more requests to other pages to grab content like rich media, and then puts all this together into an interactive display that you can use on your desktop.

So, at a high level, our computer asks for something, and the web server sends it back to us!

Throughout this demo we are going to work with the webpage for the [Data Science course at the University of Maryland by Prof. John P. Dickerson](http://www.cs.umd.edu/class/fall2018/cmsc320/). The course here at Tulane owes a big debt to John who helped us come up with much of the content for this course.

To start out with, let's try to get the contents of the course, not in our web-browser but instead just to see the text! First we'll use a Linux command line tool called [curl](https://curl.se/) to just grab the contents of the page.

In [2]:
!curl 'http://www.cs.umd.edu/class/fall2018/cmsc320/'

<!DOCTYPE html>
<html lang="en">
  <head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- The above 3 meta tags *must* come first in the head; any other head content must come *after* these tags -->
    <meta name="description" content="Introduction to Data Science">
    <meta name="author" content="John P. Dickerson">
    <link rel="icon" href="favicon.ico">

    <title>Introduction to Data Science</title>

    <!-- Bootstrap core CSS -->
    <link href="css/bootstrap.min.css" rel="stylesheet">

    <!-- IE10 viewport hack for Surface/desktop Windows 8 bug -->
    <link href="css/ie10-viewport-bug-workaround.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="css/jumbotron-narrow.css" rel="stylesheet">

    <!-- Just for debugging purposes. Don't actually copy these 2 lines! -->
    <!--[if lt IE 9]><script src="../../assets/js/ie8-re

You can see that the contents of the site is the same as if we had gone to the page and clicked the `view source` button that is available in most browsers. That is, the site returned all its information in HTML, Hyper Text Markup Language. HTML has a format and specification that is controlled by the [w3C Standard](https://html.spec.whatwg.org/multipage/) and we're currently on HTML 5, the newest version.

In short, HTML is a system of tags, each inside a bracket. So for instance, the `<body>` tag typically marks the main body of an HTML page. You can read more about the tags at the page above, for our purposes we'll be mainly focused on `<table>` tags, because that's where lots of data lives!

## Using the Requests Library

In Python we are going to generally make use of the [requests library, whose full documentation can be found at this site](https://requests.readthedocs.io/en/latest/).

In general, Requests allows you to send HTTP/1.1 requests extremely easily. There’s no need to manually add query strings to your URLs, or to form-encode your POST data. Keep-alive and HTTP connection pooling are 100% automatic, thanks to [urllib3](https://github.com/urllib3/urllib3).

Requests can do many more things than we will use it for here, but for now let's grab the webpage from John and read it into a variable.

In [3]:
import requests

r = requests.get('http://www.cs.umd.edu/class/fall2018/cmsc320/', timeout=10)
print(r.status_code)
print(r.content[:1000])

200
b'<!DOCTYPE html>\n<html lang="en">\n  <head>\n\n    <meta charset="utf-8">\n    <meta http-equiv="X-UA-Compatible" content="IE=edge">\n    <meta name="viewport" content="width=device-width, initial-scale=1">\n    <!-- The above 3 meta tags *must* come first in the head; any other head content must come *after* these tags -->\n    <meta name="description" content="Introduction to Data Science">\n    <meta name="author" content="John P. Dickerson">\n    <link rel="icon" href="favicon.ico">\n\n    <title>Introduction to Data Science</title>\n\n    <!-- Bootstrap core CSS -->\n    <link href="css/bootstrap.min.css" rel="stylesheet">\n\n    <!-- IE10 viewport hack for Surface/desktop Windows 8 bug -->\n    <link href="css/ie10-viewport-bug-workaround.css" rel="stylesheet">\n\n    <!-- Custom styles for this template -->\n    <link href="css/jumbotron-narrow.css" rel="stylesheet">\n\n    <!-- Just for debugging purposes. Don\'t actually copy these 2 lines! -->\n    <!--[if lt IE 9]><scr

## Using Beautiful Soup

So all that HTML is great but, like many things in this class, we don't want to have to read through all those tags ourselves! Imagine having to write your own parser for all that :-(

Luckily there is a great library called [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/) which can handle many of these things for us. Beautiful Soup is a Python library for pulling data out of HTML and XML files. It works with your favorite parser to provide idiomatic ways of navigating, searching, and modifying the parse tree. It commonly saves programmers hours or days of work.

Let's try this out with John's webpage and navigate through all the parts of the page. The first thing we do is get the contents of the webpage and then load it into a Beautiful Soup object, which gives us an easy way to navigate bits of the page. We can use the `prettify` specifier to parse John's webpage from above into the various tags of an HTML page.

In [4]:
from bs4 import BeautifulSoup

r = requests.get('http://www.cs.umd.edu/class/fall2018/cmsc320/')
soup = BeautifulSoup( r.content )
print(soup.prettify())

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <meta content="IE=edge" http-equiv="X-UA-Compatible"/>
  <meta content="width=device-width, initial-scale=1" name="viewport"/>
  <!-- The above 3 meta tags *must* come first in the head; any other head content must come *after* these tags -->
  <meta content="Introduction to Data Science" name="description"/>
  <meta content="John P. Dickerson" name="author"/>
  <link href="favicon.ico" rel="icon"/>
  <title>
   Introduction to Data Science
  </title>
  <!-- Bootstrap core CSS -->
  <link href="css/bootstrap.min.css" rel="stylesheet"/>
  <!-- IE10 viewport hack for Surface/desktop Windows 8 bug -->
  <link href="css/ie10-viewport-bug-workaround.css" rel="stylesheet"/>
  <!-- Custom styles for this template -->
  <link href="css/jumbotron-narrow.css" rel="stylesheet"/>
  <!-- Just for debugging purposes. Don't actually copy these 2 lines! -->
  <!--[if lt IE 9]><script src="../../assets/js/ie8-responsive-file-warning.j

More than that, soup gives us an array of commands to allow us to get the title of the page and many other things! Let's try it out.

In [5]:
# Get the webpage title tag
print(soup.title)

# Get just the string
print(soup.title.string)


<title>Introduction to Data Science</title>
Introduction to Data Science


One common thing we can do is to get all the `text` from a page, without all the markup! For John's page this includes a lot of whitespace so we'll use the [Python string manipulation function](https://docs.python.org/3.3/library/stdtypes.html?highlight=split) called `split` to split up all the whitespace and just get all the words on the page in an array, one word at a time.

In [8]:
# Only print the first 50 words to make life easier.

print(soup.get_text().split()[:50])

['Introduction', 'to', 'Data', 'Science', 'Description', 'Schedule', 'Assignments', 'CMSC320', '–', 'Fall', '2018', 'Introduction', 'to', 'Data', 'Science', 'Instructors:', 'John', 'P.', 'Dickerson', 'and', 'Prem', 'Saggar', 'TAs:', 'Eleftheria', 'Briakou,', 'Hao', 'Chen,', 'Susmija', 'Jabbireddy,', 'Harihara', 'Muralidharan,', 'Renkun', 'Ni,', 'Nischal', 'Reddy', '(UGTA),', 'Yanchao', 'Sun', 'Lectures:', 'Monday', 'and', 'Wednesday,', '2:00–3:15', 'PM,', 'ESJ', '0224', 'Description', 'of', 'Course', 'Data']


Another common thing is to take a webpage and get all the links on that page. For a hyperlink, we use the `<a href="DEST">Text</a>` tag as these are called `hyper-references` (everything was really hyper back in the 80's). Beautiful Soup has a pretty easy way to do this!

In [ ]:
# Iterate over all the <a> tags and get the target (href) element.
for link in soup.find_all('a'):
    print(link.get('href'))

Finally, perhaps the most important part for us is to get at all the tables on a particular webpage. If you go look at the webpage, you'll see several elements that look like a table. In HTML a tables always looks like this:

```
<table class="table table-striped">
<caption>Office Hours</caption>
<thead>
<tr>
<th style="width: 30%">Human</th>
<th style="width: 45%">Time</th>
<th style="width: 25%">Location</th>
</tr>
</thead>
<tbody>
<tr>
<td>Eleftheria Briakou</td>
<td>Mondays, 11:45am–1:45pm</td>
<td>AVW 1120</td>
</tr>
...
```

Where the <table> tag begins a table and closes with `</table>`. Likewise a table can have a `<caption>` as well as a table head marked `<thead>`. For each row of a table we start with `<tr>` and end that row with `</tr>` separating elements with `<td>` tags.

So we can use Beautiful Soup to grab the first table using the `find` command. Note that `find` will by default only grab the first table.


In [10]:
# Grab the first table.
soup.find("table")

<table class="table table-striped">
<caption>Office Hours</caption>
<thead>
<tr>
<th style="width: 30%">Human</th>
<th style="width: 45%">Time</th>
<th style="width: 25%">Location</th>
</tr>
</thead>
<tbody>
<tr>
<td>Eleftheria Briakou</td>
<td>Mondays, 11:45am–1:45pm</td>
<td>AVW 1120</td>
</tr>
<tr>
<td>Hao Chen</td>
<td>Fridays, 3pm–5pm</td>
<td>AVW 1120</td>
</tr>
<tr>
<td>John Dickerson</td>
<td>TBD 1hr/wk.  Also by appointment; please email John with <code>[CMSC320]</code> in the email subject line.</td>
<td>AVW 3217</td>
</tr>
<tr>
<td>Susmija Jabbireddy</td>
<td>Thursdays, 3pm–5pm</td>
<td>AVW 1120</td>
</tr>
<tr>
<td>Harihara Muralidharan</td>
<td>Mondays, 10am–12pm</td>
<td>AVW 1120</td>
</tr>
<tr>
<td>Renkun Ni</td>
<td>Wednesdays, 11:00am–1:00pm</td>
<td>AVW 1120</td>
</tr>
<tr>
<td>Nischal Reddy</td>
<td>Thursdays, 9am–10am</td>
<td>AVW 1120</td>
</tr>
<tr>
<td>Prem Saggar</td>
<td>By appointment; please email Prem with <code>[CMSC320]</code> in the email subject line.</td

Now, here's the cool part. Pandas knows how to read HTML! Specifically it knows how to read HTML tables, using the [read_html()](https://pandas.pydata.org/docs/reference/api/pandas.read_html.html) function. There are quite a few gotchas with working with HTML so for any page you want to read you'll have to make sure the table is well formatted, and that the Pandas function is working.

Note that there are more than a few **Gotcha's** in the below example.
1. `find` lazily evaluates the search, so we have to use the `str` command to force Beautiful Soup to return a string that can be passed into `read_html` like it expects.
2. If we look at the documentation for `read_html` we find that it returns a **list of tables** since there might be more than one, so here we need to simply display the first table.

In [11]:
import pandas as pd
df_t = pd.read_html(str(soup.find("table")))
display(df_t[0])

Human                                               Time  \
0     Eleftheria Briakou                            Mondays, 11:45am–1:45pm   
1               Hao Chen                                   Fridays, 3pm–5pm   
2         John Dickerson  TBD 1hr/wk. Also by appointment; please email ...   
3     Susmija Jabbireddy                                 Thursdays, 3pm–5pm   
4  Harihara Muralidharan                                 Mondays, 10am–12pm   
5              Renkun Ni                         Wednesdays, 11:00am–1:00pm   
6          Nischal Reddy                                Thursdays, 9am–10am   
7            Prem Saggar  By appointment; please email Prem with [CMSC32...   
8            Yanchao Sun                                Tuesdays, 10am–12pm   

   Location  
0  AVW 1120  
1  AVW 1120  
2  AVW 3217  
3  AVW 1120  
4  AVW 1120  
5  AVW 1120  
6  AVW 1120  
7       NaN  
8  AVW 1120

In [ ]:
df_tables = []
for t in soup.findAll("table"):
    df_t = pd.read_html(str(t))
    df_tables.append(df_t[0])

for t in df_tables:
    display(t)

And now we are close to what we need -- we can read all the tables into DataFrames and then we know how to manipulate these.

One other thing that might be useful is not necessarily loading all the tables in, but only certain ones. If you go look at the page source you'll see that the tables have a type associated with them, `<table class="table table-striped">`. If there are many different styles of table on a page, then we could pass this into Beautiful Soup to only find those tables that have that class property, as we do in the next example.

In [ ]:
print(soup.find("table", {"class": "table-striped"}))

While this isn't super useful just yet, since each table on the website isn't tagged, we can use it to search for other things. If you go look at the layout of the page you'll notice that there are `<div>` tags, which are short for page divisions, and each one of these has an `id` element. So, if wanted to find the table in the assignments section of the page only, we can do the following.

In [ ]:
df_assignments = pd.read_html(str(soup.find("div", {"id": "assignments"}).find("table")))[0]
display(df_assignments)

## Putting It All Together: Reading all the PDFs on a Webpage

Consider the following, very real, example. Let's say that you waited until the last night before the final and you need to download all the PDFs from the webpage so you can print them off really small for your cheat sheet! However, you've been cramming pizza as part of your study routine so you can't right-click the mouse any more. So we'll need to write a parser to go read the webpage, grab all the links in the lectures table, and only keep the ones that end in `.pdf` ... Easy right?

Let's take this step by step...

First we'll write a command to go grab the table we want, and extract the links from that section.

In [ ]:
soup.find("div", {"id": "schedule"}).find_all('a')

Now, we can see that there is a whole mix of different things here, and we for sure don't want them all! In fact, we really only want the ones from the `slides` column of the table, we could go that route but instead let's just pull all the links that end in `.pdf`.

In [ ]:
pdfs = []
links = soup.find("div", {"id": "schedule"}).find_all('a')
for l in links:
    # Just get the reference link
    href = l['href']
    # If it's a PDF, save it...
    if href.lower().endswith(('.pdf')):
        pdfs.append(href)
print(pdfs)

**Note:** The next cell will download all the files and save them locally, so only run it if you really want them!

So now that we have the list `pdfs`, we just need to walk through it and issue a `get` request for the data. We know this data is a PDF but our system does not. So the best way to do this is to download the file in `binary` which just gets the individual bits, and then save them.

In [ ]:
from urllib.parse import urljoin
import os

base_url = "http://www.cs.umd.edu/class/fall2018/cmsc320/"

for href in pdfs:
    urld = urljoin(base_url, href)
    print(urld)
    rd = requests.get(urld, stream=True)
    
    # Write the downloaded PDF to a file
    # Note because the href is a path we have to just get the filename!
    outfile = os.path.join("./", href.split("/")[-1])
    print("Writing: ",outfile)
    with open(outfile, 'wb') as f:
        f.write(rd.content)

### An Extra Example

So note that above we used Beautiful Soup to really get all the bits of the webpage we need. Back in the olden days, one could iterate through all the tags on the webpage to get at the same information, but it would be a lot harder. To prove this point, Karthik wrote the below code that runs through each bit of the webpage, and extracts the tables, and then keeps track of the link. This is a much more manual process but it serves to illustrate how tough this can be.

In [ ]:
from IPython.display import display, Markdown

r = requests.get( "http://www.cs.umd.edu/class/fall2018/cmsc320/" )
soup = BeautifulSoup(r.content )

tables = soup.findAll("table")

for i,t in enumerate(tables):
    # get table head for column names
    table_head = t.find("thead")
    column_names = [t.text for t in table_head.findAll("th")]
    # extract row data from table body
    table_body = t.find("tbody")
    row_tags = [r for r in table_body.findAll("tr")]
    all_rows = []
    for r_ in row_tags:
        row_data = []
        for row_item_tag in r_.findAll(["td","th"]):
            row_entry = None
            # if row item has some link , save that as well as a tuple entry in the table
            if row_item_tag.find("a"):
                row_entry = (row_item_tag.text,[x["href"] for x in row_item_tag.findAll("a")])
            else:
                row_entry = row_item_tag.text
            
            row_data.append(row_entry)
    
        all_rows.append(row_data)
    
    df_t = pd.DataFrame(all_rows,columns=column_names)
    display(Markdown(f"# Table{i+1}"))
    display(df_t)
    print("\n\n")

## Exercises

All of the exercises in this section will build on one another, you'll be attempting to get some **very messy** data from the internet and make it useable and beautiful like the data we've provided for you all semester.

You've been hired by a new space weather startup looking to disrupt the space weather reporting business. Your first project is to provide better data about the top 50 solar flares recorded so far than that shown by your competitor [SpaceWeatherLive.com](https://www.spaceweatherlive.com/en/solar-activity/top-50-solar-flares). To do this, they've pointed you to [this messy HTML page](http://cdaw.gsfc.nasa.gov/CME_list/radio/waves_type2.html) from NASA ([available here also](http://www.hcbravo.org/IntroDataSci/misc/waves_type2.html)) where you can get the extra data your startup is going to post in your new spiffy site.

Of course, you don't have access to the raw data for either of these two tables, so as an enterprising data scientist you will scrape this information directly from each HTML page using all the great tools available to you in Python. By the way, you should read up a bit on [Solar Flares](https://en.wikipedia.org/wiki/Solar_flare), [coronal mass ejections](https://www.spaceweatherlive.com/en/help/what-is-a-coronal-mass-ejection-cme), [the solar flare alphabet soup](http://spaceweather.com/glossary/flareclasses.html), [the scary storms of Halloween 2003](http://www.nasa.gov/topics/solarsystem/features/halloween_storms.html), and [sickening solar flares](https://science.nasa.gov/science-news/science-at-nasa/2005/27jan_solarflares).

In this Lab we'll first get the data into a format we can sort of use, and next lab we'll spend some time cleaning the data and attempting to match them up with one another.

### Exercise 1

The first thing we need to do is download the webpage into a variable. So write a command with the `requests` module to get the webpage at <https://www.spaceweatherlive.com/en/solar-activity/top-50-solar-flares>. Once you issue the command, print the status code of the webpage along with the text, is this what you expected?

A helpful reference for this part is the [documentation for the requests module](https://requests.readthedocs.io/en/latest/) along with the documentation for [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/).

Find the code at the [Mozilla Foundation Webpage](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) and briefly explain the problem.

In [23]:
# Your Code Here.
r = requests.get('https://www.spaceweatherlive.com/en/solar-activity/top-50-solar-flares', timeout=10)
print(f'Status Code: {r.status_code}')
r.text

#soup = BeautifulSoup(r.content)
#print(soup.prettify())

Status Code: 403


'<!doctype html> <html lang="en"> <head> <meta charset="utf-8"> <meta http-equiv="x-ua-compatible" content="ie=edge"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <title>StackPath</title> <style> *{box-sizing: border-box;}*::before, *::after{box-sizing: border-box;}*:focus{outline: none;}html{-moz-osx-font-smoothing: grayscale;-ms-overflow-style: -ms-autohiding-scrollbar;-webkit-font-smoothing: antialiased;font-size: 16px;}body{background-color: #fff;color: #545963;font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif;font-weight: 400;line-height: 1.6;margin: 0;min-width: 20rem;}h1, p{margin: 0;padding: 0;}h1{font-size: 2rem;font-weight: 400;line-height: 1;}h1:not(:last-child){margin-bottom: 3rem;}p{font-size: 1rem;}p:not(:last-child){margin-bottom: 1.5rem;}a{color: #0934a0;text-decoration: none;}a:hover{color: #0934a0;text-decoration: under

**Written Answer:** Fill in your answer here.

### Exercise 2

Welp, looks like they don't want us to scrape their webpage. There are a few tricks we can use to get around this, but shhhh! don't tell anyone.

We can add a few things to our [Requests Headder](https://requests.readthedocs.io/en/master/user/quickstart/#custom-headers), [Maybe try something like this](https://stackoverflow.com/questions/27652543/how-to-use-python-requests-to-fake-a-browser-visit), to trick the server into thinking you're human, but don't tell anyone :-).

After reading those pages, write another request command below that successfully gets the webpage, print out the status code and the first 500 characters of the webpage.

In [24]:
# Your Code Here.
import requests

url = 'https://www.spaceweatherlive.com/en/solar-activity/top-50-solar-flares'
headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'}
r = requests.get(url, headers=headers)

print(f'Status Code: {r.status_code}')
soup = BeautifulSoup(r.content)
print(soup.prettify()[:500])

Status Code: 200
<!DOCTYPE html>
<html lang="en">
 <head>
  <title>
   Top 50 solar flares | Solar activity | SpaceWeatherLive.com
  </title>
  <meta charset="utf-8"/>
  <meta content="index, follow, max-snippet:-1, max-image-preview:large, max-video-preview:-1" name="robots"/>
  <meta content="On this page you will find an overview of the strongest solar flares since June 1996 together with links to more information in our archive and a v..." name="description"/>
  <meta content="SpaceWeatherLive, Live, Aurora,


### Exercise 3

Well now that we have all that text, we need to figure out how to get what we want: just the data out of the table. In the below cell write code that takes the response from the website and uses the [Beautiful Soup find Function](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find) to extract the only table on the page. Print the first 600 characters of the table.

In [25]:
# Your Code Here.
str(soup.find('table'))[:600]

'<table class="table table-striped"><thead><tr><th></th><th></th><th></th><th>Region</th><th>Start</th><th>Maximum</th><th>End</th><th></th></tr></thead><tbody><tr><td>1</td><td><span class="flare_small Xclass">X28+</span></td><td>2003/11/04</td><td><a class="btn btn-SWL RegionBGD" href="/en/solar-activity/region/10486.html">0486</a></td><td>19:29</td><td>19:53</td><td>20:06</td><td><div class="btn-group"><a class="btn btn-xs btn-outline-dark" data-fancybox="" href="/images/Archief/2003/Flares/200311041953X28.0.mp4"><i aria-hidden="true" class="fa-regular fa-play-circle"></i> Movie</a><a class='

### Exercise 4

Well, there's the table! Now what do we do with it? If you look again at the website you'll see that the table looks and awful awful lot like a table we'd use in Pandas. Well, luckily, Pandas has a great function [read_html](https://pandas.pydata.org/docs/reference/api/pandas.read_html.html_) which works a whole lot like the `read_csv` function we've been using all class. In the below use the `read_html` function to load all this into an dataframe, print the first 10 rows of the table.

**Gotcha Warning:** Read the docs on `read_html` closely -- what is the return type? Make sure you end up with a dataframe!

In [26]:
# Your Code Here.
df_tab = pd.read_html(str(soup.find('table')))
df_tab = df_tab[0]
df_tab[:10]

Unnamed: 0 Unnamed: 1  Unnamed: 2  Region  Start Maximum    End  \
0           1       X28+  2003/11/04     486  19:29   19:53  20:06   
1           2       X20+  2001/04/02    9393  21:32   21:51  22:03   
2           3     X17.2+  2003/10/28     486  09:51   11:10  11:24   
3           4       X17+  2005/09/07     808  17:17   17:40  18:03   
4           5      X14.4  2001/04/15    9415  13:19   13:50  13:55   
5           6        X10  2003/10/29     486  20:37   20:49  21:01   
6           7       X9.4  1997/11/06    8100  11:49   11:55  12:01   
7           8       X9.3  2017/09/06    2673  11:53   12:02  12:10   
8           9         X9  2006/12/05     930  10:18   10:35  10:45   
9          10       X8.3  2003/11/02     486  17:03   17:25  17:39   

          Unnamed: 7  
0  MovieView archive  
1  MovieView archive  
2  MovieView archive  
3  MovieView archive  
4  MovieView archive  
5  MovieView archive  
6  MovieView archive  
7  MovieView archive  
8  MovieView archive  
9  MovieView archive

### Exercise 5

Okay, now we've got a table but it's not very clean... or tidy. This is the hardest part of the whole lab! You'll need to do several steps for this part, all at once.

1. Assign reasonable names to each of the columns of the dataframe
2. Drop the last column of the dataframe, we don't need it.
3. Check the `Region` column. Make sure to set any weird codes to NaN's (though there might not be any, you should check)!
4. The dates and times aren't that useful as is, we need them to be one datetime object so we can do math on them.
5. Re-arrange the columns to be pretty. 

When you're done your table should look like the below. Note that this is a tough problem, you'll need to do several steps to achieve this one! Make sure to check the dTypes of things as you go and **Hint** you might want to check out the [Pandas to_datetime function](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html), it's the easiest way to accomplish this! 

In [27]:
# Your Code Here.

# Rename columns and drop the last column
df_tab = df_tab.rename(columns={'Unnamed: 0': 'rank', 'Region': 'region', 'Unnamed: 1': 'x_class', 'Unnamed: 2': 'Year', 'Start': 'start_datetime', 'Maximum': 'max_datetime', 'End': 'end_datetime'})
df_tab = df_tab.drop(['Unnamed: 7'], axis=1)

# Create datetime columns so we can do math on them
df_tab['start_datetime'] = pd.to_datetime(df_tab['Year'] + ' ' + df_tab['start_datetime'])
df_tab['max_datetime'] = pd.to_datetime(df_tab['Year'] + ' ' + df_tab['max_datetime'])
df_tab['end_datetime'] = pd.to_datetime(df_tab['Year'] + ' ' + df_tab['end_datetime'])
df_tab = df_tab.drop(['Year'], axis=1)

# Rearrange columns to put 'region' at the end
new_cols = ['rank', 'x_class', 'start_datetime', 'max_datetime', 'end_datetime', 'region']
df_tab = df_tab.reindex(columns=new_cols)


**In The Cell Below:** Make sure you display both the table and the final dtypes in the cell below.

In [28]:
# Your Code Here.
print(f'dtypes: {df_tab.dtypes}')
df_tab

dtypes: rank                       int64
x_class                   object
start_datetime    datetime64[ns]
max_datetime      datetime64[ns]
end_datetime      datetime64[ns]
region                     int64
dtype: object


rank x_class      start_datetime        max_datetime        end_datetime  \
0      1    X28+ 2003-11-04 19:29:00 2003-11-04 19:53:00 2003-11-04 20:06:00   
1      2    X20+ 2001-04-02 21:32:00 2001-04-02 21:51:00 2001-04-02 22:03:00   
2      3  X17.2+ 2003-10-28 09:51:00 2003-10-28 11:10:00 2003-10-28 11:24:00   
3      4    X17+ 2005-09-07 17:17:00 2005-09-07 17:40:00 2005-09-07 18:03:00   
4      5   X14.4 2001-04-15 13:19:00 2001-04-15 13:50:00 2001-04-15 13:55:00   
5      6     X10 2003-10-29 20:37:00 2003-10-29 20:49:00 2003-10-29 21:01:00   
6      7    X9.4 1997-11-06 11:49:00 1997-11-06 11:55:00 1997-11-06 12:01:00   
7      8    X9.3 2017-09-06 11:53:00 2017-09-06 12:02:00 2017-09-06 12:10:00   
8      9      X9 2006-12-05 10:18:00 2006-12-05 10:35:00 2006-12-05 10:45:00   
9     10    X8.3 2003-11-02 17:03:00 2003-11-02 17:25:00 2003-11-02 17:39:00   
10    11    X8.2 2017-09-10 15:35:00 2017-09-10 16:06:00 2017-09-10 16:31:00   
11    12    X7.1 2005-01-20 06:36:00 2005-01-20 07:01:00 2005-01-20 07:26:00   
12    13    X6.9 2011-08-09 07:48:00 2011-08-09 08:05:00 2011-08-09 08:08:00   
13    14    X6.5 2006-12-06 18:29:00 2006-12-06 18:47:00 2006-12-06 19:00:00   
14    15    X6.2 2005-09-09 19:13:00 2005-09-09 20:04:00 2005-09-09 20:36:00   
15    16    X6.2 2001-12-13 14:20:00 2001-12-13 14:30:00 2001-12-13 14:35:00   
16    17    X5.7 2000-07-14 10:03:00 2000-07-14 10:24:00 2000-07-14 10:43:00   
17    18    X5.6 2001-04-06 19:10:00 2001-04-06 19:21:00 2001-04-06 19:31:00   
18    19    X5.4 2012-03-07 00:02:00 2012-03-07 00:24:00 2012-03-07 00:40:00   
19    20    X5.4 2005-09-08 20:52:00 2005-09-08 21:06:00 2005-09-08 21:17:00   
20    21    X5.4 2003-10-23 08:19:00 2003-10-23 08:35:00 2003-10-23 08:49:00   
21    22    X5.3 2001-08-25 16:23:00 2001-08-25 16:45:00 2001-08-25 17:04:00   
22    23    X4.9 2014-02-25 00:39:00 2014-02-25 00:49:00 2014-02-25 01:03:00   
23    24    X4.9 1998-08-18 22:10:00 1998-08-18 22:19:00 1998-08-18 22:28:00   
24    25    X4.8 2002-07-23 00:18:00 2002-07-23 00:35:00 2002-07-23 00:47:00   
25    26      X4 2000-11-26 16:34:00 2000-11-26 16:48:00 2000-11-26 16:56:00   
26    27    X3.9 2003-11-03 09:43:00 2003-11-03 09:55:00 2003-11-03 10:19:00   
27    28    X3.9 1998-08-19 21:35:00 1998-08-19 21:45:00 1998-08-19 21:50:00   
28    29    X3.8 2005-01-17 06:59:00 2005-01-17 09:52:00 2005-01-17 10:07:00   
29    30    X3.7 1998-11-22 06:30:00 1998-11-22 06:42:00 1998-11-22 06:49:00   
30    31    X3.6 2005-09-09 09:42:00 2005-09-09 09:59:00 2005-09-09 10:08:00   
31    32    X3.6 2004-07-16 13:49:00 2004-07-16 13:55:00 2004-07-16 14:01:00   
32    33    X3.6 2003-05-28 00:17:00 2003-05-28 00:27:00 2003-05-28 00:39:00   
33    34    X3.4 2006-12-13 02:14:00 2006-12-13 02:40:00 2006-12-13 02:57:00   
34    35    X3.4 2001-12-28 20:02:00 2001-12-28 20:45:00 2001-12-28 21:32:00   
35    36    X3.3 2013-11-05 22:07:00 2013-11-05 22:12:00 2013-11-05 22:15:00   
36    37    X3.3 2002-07-20 21:04:00 2002-07-20 21:30:00 2002-07-20 21:54:00   
37    38    X3.3 1998-11-28 04:54:00 1998-11-28 05:52:00 1998-11-28 06:13:00   
38    39    X3.2 2013-05-14 00:00:00 2013-05-14 01:11:00 2013-05-14 01:20:00   
39    40    X3.1 2014-10-24 21:07:00 2014-10-24 21:41:00 2014-10-24 22:13:00   
40    41    X3.1 2002-08-24 00:49:00 2002-08-24 01:12:00 2002-08-24 01:31:00   
41    42      X3 2002-07-15 19:59:00 2002-07-15 20:08:00 2002-07-15 20:14:00   
42    43    X2.8 2013-05-13 15:48:00 2013-05-13 16:05:00 2013-05-13 16:16:00   
43    44    X2.8 2001-12-11 07:58:00 2001-12-11 08:08:00 2001-12-11 08:14:00   
44    45    X2.8 1998-08-18 08:14:00 1998-08-18 08:24:00 1998-08-18 08:32:00   
45    46    X2.7 2015-05-05 22:05:00 2015-05-05 22:11:00 2015-05-05 22:15:00   
46    47    X2.7 2003-11-03 01:09:00 2003-11-03 01:30:00 2003-11-03 01:45:00   
47    48    X2.7 1998-05-06 07:58:00 1998-05-06 08:09:00 1998-05-06 08:20:00   
48    49    X2.6 2005-01-15 22:25:00 2005-01-15 23:02:00 2005-01-15 23:31:00   
49  

Your table should look like this:

|    |   rank | x_class   | start_datetime      | max_datetime        | end_datetime        |   region |
|---:|-------:|:----------|:--------------------|:--------------------|:--------------------|---------:|
|  0 |      1 | X28+      | 2003-11-04 19:29:00 | 2003-11-04 19:53:00 | 2003-11-04 20:06:00 |      486 |
|  1 |      2 | X20+      | 2001-04-02 21:32:00 | 2001-04-02 21:51:00 | 2001-04-02 22:03:00 |     9393 |
|  2 |      3 | X17.2+    | 2003-10-28 09:51:00 | 2003-10-28 11:10:00 | 2003-10-28 11:24:00 |      486 |
|  3 |      4 | X17+      | 2005-09-07 17:17:00 | 2005-09-07 17:40:00 | 2005-09-07 18:03:00 |      808 |
|  4 |      5 | X14.4     | 2001-04-15 13:19:00 | 2001-04-15 13:50:00 | 2001-04-15 13:55:00 |     9415 |
|  5 |      6 | X10       | 2003-10-29 20:37:00 | 2003-10-29 20:49:00 | 2003-10-29 21:01:00 |      486 |
|  6 |      7 | X9.4      | 1997-11-06 11:49:00 | 1997-11-06 11:55:00 | 1997-11-06 12:01:00 |     8100 |
|  7 |      8 | X9.3      | 2017-09-06 11:53:00 | 2017-09-06 12:02:00 | 2017-09-06 12:10:00 |     2673 |
|  8 |      9 | X9        | 2006-12-05 10:18:00 | 2006-12-05 10:35:00 | 2006-12-05 10:45:00 |      930 |
|  9 |     10 | X8.3      | 2003-11-02 17:03:00 | 2003-11-02 17:25:00 | 2003-11-02 17:39:00 |      486 |

Congratulations, you've completed the first step of a much larger project involving data integration, and scraping multiple websites we used to do as part of a more advanced version of this course. If you're itching for more work (not bonus) you can [see the full assignment here](https://github.com/nmattei/cmps3160/blob/fall2021/_projects/Project1.md).